### PR TITLE
[rllib] Replace ray.get() with ray_get_and_free() to optimize memory usage

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

When ray.get() is called in RLlib, the result should be freed immediately. This prevents object store memory from growing unnecessarily.

## Related issue number

(many previous issues)

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by ericl** [comment]
@robertnishihara @guoyuhong do you see any performance issues with calling ray.internal.free up to thousands up to times a second here?

**Comment by guoyuhong** [comment]
@ericl Maybe you need to do some tests. `ray.internal.free` will spread the message to all raylet nodes. In our internal code, we do not call `ray.internal.free` for each object. We will collect the object ids and send them by batch.

**Comment by ericl** [comment]
Hm yeah, that might be a high overhead if we broadcast to all nodes for each object created. I could also batch the objects and periodically flush them.

**Comment by ericl** [comment]
jenkins retest this please

**Comment by ericl** [comment]
cc @joneswong

File: python/ray/rllib/utils/memory.py [hunk 1]
```diff
@@ -3,6 +3,46 @@
...
+    global _to_free
+
+    result = ray.get(object_ids)
+    if type(object_ids) is not list:
+        object_ids = [object_ids]
+    _to_free.extend(object_ids)
+
+    now = time.time()
+    if (len(_to_free) > MAX_FREE_QUEUE_SIZE
+            or now - _last_free_time > FREE_DELAY_S):
```
Comment (ericl) [inline python/ray/rllib/utils/memory.py]: batch the free calls to avoid too much overheads

**Comment by ericl** [comment]
@richardliaw  ?

File: python/ray/internal/internal_api.py [hunk 2]
```diff
@@ -30,6 +30,9 @@ def free(object_ids, local_only=False, delete_creating_tasks=False):
     """
     worker = ray.worker.get_global_worker()
 
+    if ray.worker._mode() == ray.worker.LOCAL_MODE:
```
Comment (richardliaw) [inline python/ray/internal/internal_api.py]: oh this is nice

**Review by richardliaw — APPROVED** [review]

**Pull request merged**